File: programme/binary-search/binary-search.js
// Write a programme for binary search
const binarySearch = (arr, target) => {
  let low = 0;
  let high = arr.length - 1;
  while (low <= high) {
    let mid = Math.floor((low + high) / 2);
    if (arr[mid] === target) {
      return mid;
    } else if (arr[mid] <= target) {
      low = mid + 1;
    } else if (arr[mid] >= target) {
      high = mid - 1;
    }
  }
  return -1;
};

const arr = [10, 20, 30, 40, 50];
const target = 30;

console.log("Input  : ", arr);
console.log("Target : ", target);
console.log("Output : ", binarySearch(arr, target));

File: programme/binary-search/binary-search.jl
# Write a programme for binary search

function binary_search(list, query; rev=false, lt=<, by=identity)
    if issorted(list) || issorted(list; rev=true)
        low = !rev ? 1 : length(list)
        high = !rev ? length(list) : 1
        middle(l, h) = round(Int, (l + h) // 2)
        query = by(query)

        while !rev ? low <= high : high <= low
            mid = middle(low, high)
            by(list[mid]) === query && return mid:mid
            if lt(by(list[mid]), query)
                low = !rev ? mid + 1 : mid - 1
            else
                high = !rev ? mid - 1 : mid + 1
            end
        end
        return !rev ? (low:high) : (high:low)

    else
        throw(error("List not sorted, unable to search value"))
    end
end

array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
target = 9

println("Input  : ", array)
println("Target : ", target)
println("Output : ", binary_search(array, target))
